### PR TITLE
Add copy_if_kt

### PR DESCRIPTION
**Pull request by AidanBeltonS**
Adds a copy_if kernel based on scan_kt.

### CODE REVIEW

**Review by joeatodd — CHANGES_REQUESTED** [review]
The core of the implementation is sound and seems correct. I think just the use of atomics instead of a barrier/broadcast is causing issues.

Aside from that correctness issue, just:
 - some unused variables & commented code
 - logic in the tests

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 1]
```diff
@@ -606,6 +606,210 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
     }
 }
 
+// Load function to try and get some PVC perf w/ coalesced
+template <typename Tp, typename _InRange>
+inline Tp load(sycl::sub_group sg, _InRange src, size_t i, size_t wg_stride, size_t tile_id) {
+    // if constexpr (std::is_arithmetic_v<Tp>) {
+    //   return sg.load(src.begin() + i / SUBGROUP_SIZE + wg_stride * tile_id);
+    // }
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: It is worth pursuing but for other reasons ill leave it for now. Iv just made this inline logic

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 2]
```diff
@@ -606,6 +606,210 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+                  wg_copy_if_values[count + wg_count.load()] = val;
+
+                if (wg_local_id == (wgsize - 1))
+                  wg_count += (count + satisfies_pred);
+                sycl::group_barrier(group);
+              }
+            }
+
+            // Phase 2: Global scan across wg_count
+            auto local_sum = wg_count.load();
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 3]
```diff
@@ -606,6 +606,210 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+            auto in_begin = tile_vals.get_pointer();
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Removed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 4]
```diff
@@ -606,6 +606,210 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+                size_t satisfies_pred = pred(val);
+                //size_t satisfies_pred = 0;
+                size_t count = sycl::exclusive_scan_over_group(group, satisfies_pred, sycl::plus<size_t>());
+
+                if (satisfies_pred)
+                  wg_copy_if_values[count + wg_count.load()] = val;
+
+                if (wg_local_id == (wgsize - 1))
+                  wg_count += (count + satisfies_pred);
+                sycl::group_barrier(group);
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 5]
```diff
@@ -606,6 +606,210 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+              #pragma unroll
+              for (size_t i = wg_local_id; i < elems_in_tile; i += wgsize) {
+                _Type val = load<_Type>(sg, __in_rng, i, elems_in_tile, tile_id);
+
+                size_t satisfies_pred = pred(val);
+                //size_t satisfies_pred = 0;
+                size_t count = sycl::exclusive_scan_over_group(group, satisfies_pred, sycl::plus<size_t>());
+
+                if (satisfies_pred)
+                  wg_copy_if_values[count + wg_count.load()] = val;
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 6]
```diff
@@ -606,6 +606,210 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+
+            // Phase 1: Create wg_count and construct in-order wg_copy_if_values
+            if ((tile_id + 1) * elems_in_tile <= n) {
+              #pragma unroll
+              for (size_t i = wg_local_id; i < elems_in_tile; i += wgsize) {
+                _Type val = load<_Type>(sg, __in_rng, i, elems_in_tile, tile_id);
+
+                size_t satisfies_pred = pred(val);
+                //size_t satisfies_pred = 0;
+                size_t count = sycl::exclusive_scan_over_group(group, satisfies_pred, sycl::plus<size_t>());
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Iv taken this approach, code is a bit cleaner

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 7]
```diff
@@ -606,6 +606,210 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+            }
+
+            // Global load into local
+            auto wg_current_offset = (tile_id * elems_in_tile);
+            auto wg_local_memory_size = elems_in_tile;
+
+            // Must be a better way to init atomics
+            l_wg_count[0] = 0;
+            sycl::group_barrier(group);
+            sycl::atomic_ref<size_t, sycl::memory_order::acq_rel, sycl::memory_scope::work_group, sycl::access::address_space::local_space> wg_count(l_wg_count[0]);
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: I have moved away from atomics and am using group_broadcast instead

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 8]
```diff
@@ -606,6 +606,210 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+                                     status_flags_begin[id] = id < _LookbackScanMemory::padding
+                                                                  ? _LookbackScanMemory::OUT_OF_BOUNDS
+                                                                  : _LookbackScanMemory::NOT_READY;
+                                 if (id == num_elements)
+                                     tile_id_begin[0] = 0;
+                             });
+        });
+
+    auto event = __queue.submit([&](sycl::handler& hdl) {
+        auto wg_copy_if_values = sycl::local_accessor<_Type, 1>(sycl::range<1>{elems_per_workitem*wgsize}, hdl);
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Done

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 9]
```diff
@@ -606,6 +606,210 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+        auto l_wg_count = sycl::local_accessor<size_t, 1>(sycl::range<1>{1}, hdl);
+
+        auto tile_id_lacc = sycl::local_accessor<std::uint32_t, 1>(sycl::range<1>{1}, hdl);
+        auto tile_vals = sycl::local_accessor<_Type, 1>(sycl::range<1>{elems_in_tile}, hdl);
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Done

File: test/parallel_api/numeric/numeric.ops/copy_if_kt.pass.cpp [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        sycl::free(in_ptr, q);
+        sycl::free(out_ptr, q);
+        sycl::free(out_num, q);
+    }
+
+    return !all_passed;
+}
+
+int main() {
+    bool all_passed;
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply test/parallel_api/numeric/numeric.ops/copy_if_kt.pass.cpp]: Sorted out the logic

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 11]
```diff
@@ -719,6 +724,264 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+        auto wg_copy_if_values = sycl::local_accessor<_Type, 1>(sycl::range<1>{elems_in_tile}, hdl);
+
+        oneapi::dpl::__ranges::__require_access(hdl, __in_rng, __out_rng, __num_rng);
+        hdl.parallel_for(sycl::nd_range<1>(num_workitems, wgsize), [=](const sycl::nd_item<1>& item)  [[intel::reqd_sub_group_size(SUBGROUP_SIZE)]] {
+            auto group = item.get_group();
+            auto wg_local_id = item.get_local_id(0);
+            constexpr ::std::uint32_t stride = wgsize;                 
+                                                            
+            // Global load into local
+            auto wg_current_offset = 0;
```
Comment (joeatodd) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:755]: Unused

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 12]
```diff
@@ -719,6 +724,264 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+
+    const ::std::size_t n = __in_rng.size();
+
+    constexpr ::std::size_t wgsize = _KernelParam::workgroup_size;
+    constexpr ::std::size_t elems_per_workitem = _KernelParam::elems_per_workitem;
+
+    // Avoid non_uniform n by padding up to a multiple of wgsize
+    constexpr std::uint32_t elems_in_tile = wgsize * elems_per_workitem;
+    ::std::size_t num_wgs = oneapi::dpl::__internal::__dpl_ceiling_div(n, elems_in_tile);
+    ::std::size_t num_workitems = num_wgs * wgsize;
```
Comment (joeatodd) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:743]: Here you should assert that there's only 1 wg.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 13]
```diff
@@ -719,6 +724,264 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+
+    event.wait();
+}
+
+template <typename _KernelParam, typename _UseAtomic64, typename _UseDynamicTileID, typename _InRange, typename _OutRange, typename _NumSelectedRange, typename _UnaryPredicate>
+void
+single_pass_copy_if_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __out_rng, _NumSelectedRange __num_rng, _UnaryPredicate pred)
+{
+    using _Type = oneapi::dpl::__internal::__value_t<_InRange>;
+    using _SizeT = uint64_t;
```
Comment (joeatodd) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:811]: Don't worry about this for now, but could be a ticket: For cases up to 2^30, we can perform the lookback with a single 32-bit int (2 bits for flags). Could template and runtime dispatch depending on `n`.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h [hunk 14]
```diff
@@ -719,6 +724,264 @@ single_pass_inclusive_scan(sycl::queue __queue, _InIterator __in_begin, _InItera
...
+
+        auto tile_id_lacc = sycl::local_accessor<std::uint32_t, 1>(sycl::range<1>{1}, hdl);
+        hdl.depends_on(fill_event);
+
+        oneapi::dpl::__ranges::__require_access(hdl, __in_rng, __out_rng, __num_rng);
+        hdl.parallel_for(sycl::nd_range<1>(num_workitems, wgsize), [=](const sycl::nd_item<1>& item)  [[intel::reqd_sub_group_size(SUBGROUP_SIZE)]] {
+            auto group = item.get_group();
+            auto wg_local_id = item.get_local_id(0);
+            auto sg = item.get_sub_group();
+            constexpr ::std::uint32_t stride = wgsize;
```
Comment (joeatodd) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:852]: Formatting around here. extra spaces

**Review by joeatodd — APPROVED** [review]
Really nice work @AidanBeltonS  :tada:

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h (hunk 11 above)
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Removed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h (hunk 12 above)
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:746]: Added assert

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h (hunk 14 above)
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h]: Iv clang-formatted the commits

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h (hunk 13 above)
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h:825]: Made ticket

**Pull request merged**